File: lib/redbook/protocol-governance.ts
import {createPublicKey,verify as verifySignature} from 'crypto';
import {canonicalHash,canonicalize} from '@/lib/server/hash';
import {
 PROTOCOL_CHANGE_DOMAIN,
 protocolChangeProofSchema,
 protocolGovernancePolicySchema,
 type ProtocolChangeProof,
 type ProtocolGovernancePolicy,
 type ProtocolState,
} from './schema/protocol-governance';

function governanceMemberEligibleAtHeight(member:ProtocolGovernancePolicy['members'][number],height:number){
 return member.state==='ACTIVE'&&member.validFromHeight<=height&&(member.validUntilHeight===null||member.validUntilHeight>height);
}

export function canonicalProtocolGovernancePolicy(input:ProtocolGovernancePolicy|unknown){
 const policy=protocolGovernancePolicySchema.parse(input);
 return{
  ...policy,
  scope:[...policy.scope].sort(),
  members:[...policy.members].sort((a,b)=>a.authorityMemberId.localeCompare(b.authorityMemberId)||a.keyId.localeCompare(b.keyId)),
 };
}

export function protocolGovernancePolicyHash(input:ProtocolGovernancePolicy|unknown){return canonicalHash(canonicalProtocolGovernancePolicy(input));}
export function protocolStateHash(state:ProtocolState|unknown){return canonicalHash(state);}
export function protocolChangeManifestHash(manifest:ProtocolChangeProof['changeManifest']){return canonicalHash(manifest);}

export function requiredProtocolAuthority(policy:ProtocolGovernancePolicy,eligibleCount:number){
 if(!Number.isInteger(eligibleCount)||eligibleCount<1)throw new Error('Protocol governance requires at least one eligible authority member');
 let floor=1;
 if(policy.thresholdClass==='DUAL'||policy.thresholdClass==='MULTI_PARTY_HIGH_ASSURANCE')floor=2;
 if(policy.thresholdClass==='MAJORITY')floor=Math.floor(eligibleCount/2)+1;
 if(policy.thresholdClass==='SUPERMAJORITY')floor=Math.floor((2*eligibleCount)/3)+1;
 return Math.max(policy.threshold,floor);
}

function requireScope(policy:ProtocolGovernancePolicy,scope:ProtocolGovernancePolicy['scope'][number]){
 if(!policy.scope.includes(scope))throw new Error(`Protocol governance policy lacks ${scope} authority`);
}

function assertStateScopes(proof:ProtocolChangeProof){
 const {governancePolicy:policy,previousProtocolState:previous,nextProtocolState:next}=proof;
 requireScope(policy,'PROTOCOL_VERSION');
 if(previous.consensusRulesHash!==next.consensusRulesHash)requireScope(policy,'CONSENSUS_RULES');
 if(previous.canonicalSchemaVersion!==next.canonicalSchemaVersion)requireScope(policy,'CANONICAL_SCHEMA');
 if(previous.resourcePolicyHash!==next.resourcePolicyHash)requireScope(policy,'RESOURCE_POLICY');
}

export function verifyProtocolChangeProof(args:{
 proof:unknown;
 expectedChainId:string;
 expectedGovernancePolicyHash:string;
 expectedPreviousProtocolStateHash:string;
}){
 const proof=protocolChangeProofSchema.parse(args.proof);
 const {action,governancePolicy:policy,changeManifest:manifest,previousProtocolState:previous,nextProtocolState:next}=proof;
 if(action.chainId!==args.expectedChainId||policy.chainId!==args.expectedChainId||manifest.chainId!==args.expectedChainId||previous.chainId!==args.expectedChainId||next.chainId!==args.expectedChainId)throw new Error('Protocol governance chainId mismatch');
 if(action.activationHeight<=action.approvedAtHeight)throw new Error('Consensus-affecting protocol change must activate after approval height');
 if(policy.status!=='ACTIVE'||policy.effectiveFromHeight>action.approvedAtHeight||(policy.effectiveUntilHeight!==null&&policy.effectiveUntilHeight<=action.approvedAtHeight))throw new Error('PROTOCOL governance policy was not ACTIVE at approval height');
 const computedPolicyHash=protocolGovernancePolicyHash(policy);
 if(computedPolicyHash!==args.expectedGovernancePolicyHash||action.governancePolicyHash!==computedPolicyHash)throw new Error('Protocol governance policy hash does not match independently trusted policy');
 if(action.fromProtocolVersion!==previous.protocolVersion||action.toProtocolVersion!==next.protocolVersion||manifest.fromProtocolVersion!==previous.protocolVersion||manifest.toProtocolVersion!==next.protocolVersion)throw new Error('Protocol version transition is not consistently bound across action, manifest and states');
 if(previous.protocolVersion===next.protocolVersion)throw new Error('Protocol activation does not change protocolVersion');
 if(previous.activeFromHeight>action.activationHeight-1)throw new Error('Previous protocol state was not active before transition height');
 if(next.activeFromHeight!==action.activationHeight)throw new Error('Next protocol state must activate exactly at governed activationHeight');
 const previousHash=protocolStateHash(previous);const nextHash=protocolStateHash(next);
 if(previousHash!==args.expectedPreviousProtocolStateHash||action.previousProtocolStateHash!==previousHash)throw new Error('Previous protocol state hash does not match independently trusted state');
 if(action.nextProtocolStateHash!==nextHash)throw new Error('Next protocol state hash mismatch');
 const changeHash=protocolChangeManifestHash(manifest);
 if(action.changeHash!==changeHash)throw new Error('Protocol change manifest hash mismatch');
 assertStateScopes(proof);

 const eligible=policy.members.filter(member=>governanceMemberEligibleAtHeight(member,action.approvedAtHeight));
 const required=requiredProtocolAuthority(policy,eligible.length);
 if(required>eligible.length)throw new Error(`Protocol governance threshold ${required} exceeds ${eligible.length} eligible authority members`);
 const byId=new Map(eligible.map(member=>[member.authorityMemberId,member]));
 const payload=Buffer.from(canonicalize(action),'utf8');const validSigners:string[]=[];
 for(const signature of proof.signatures){
  const member=byId.get(signature.authorityMemberId);
  if(!member||signature.keyId!==member.keyId||signature.algorithm!=='Ed25519'||signature.domain!==PROTOCOL_CHANGE_DOMAIN)continue;
  const publicKey=createPublicKey({key:Buffer.from(member.publicKeyDerB64,'base64'),format:'der',type:'spki'});
  if(verifySignature(null,payload,publicKey,Buffer.from(signature.signatureB64,'base64')))validSigners.push(member.authorityMemberId);
 }
 const uniqueValid=[...new Set(validSigners)].sort();
 if(uniqueValid.length<required)throw new Error(`Protocol governance threshold not met: ${uniqueValid.length}/${eligible.length}; ${required} required`);
 return{
  valid:true as const,
  actionId:action.actionId,
  approvedAtHeight:action.approvedAtHeight,
  activationHeight:action.activationHeight,
  fromProtocolVersion:previous.protocolVersion,
  toProtocolVersion:next.protocolVersion,
  governancePolicyHash:computedPolicyHash,
  changeHash,
  previousProtocolStateHash:previousHash,
  nextProtocolStateHash:nextHash,
  eligibleAuthorityCount:eligible.length,
  requiredAuthority:required,
  validSigners:uniqueValid,
  trustedNextProtocolState:{activationHeight:action.activationHeight,protocolVersion:next.protocolVersion,stateHash:nextHash},
 };
}

File: tools/portable-validator/protocol_governance.go
package main

import (
	"crypto/ed25519"
	"crypto/x509"
	"encoding/base64"
	"errors"
	"fmt"
	"sort"
)

const (
	protocolGovernanceProfile      = "STRATUM-PROTOCOL-GOVERNANCE/1"
	protocolChangeVersion          = "STRATUM-PROTOCOL-CHANGE/1"
	protocolChangeDomain           = "STRATUM/PROTOCOL_CHANGE/1"
	protocolChangeManifestVersion  = "STRATUM-PROTOCOL-CHANGE-MANIFEST/1"
	protocolStateProfile           = "STRATUM-PROTOCOL-STATE/1"
)

type ProtocolGovernanceMember struct {
	AuthorityMemberID string `json:"authorityMemberId"`
	KeyID             string `json:"keyId"`
	Purpose           string `json:"purpose"`
	Algorithm         string `json:"algorithm"`
	PublicKeyDerB64   string `json:"publicKeyDerB64"`
	State             string `json:"state"`
	ValidFromHeight   int64  `json:"validFromHeight"`
	ValidUntilHeight  *int64 `json:"validUntilHeight"`
}

type ProtocolGovernancePolicy struct {
	Profile              string                     `json:"profile"`
	ChainID              string                     `json:"chainId"`
	AuthorityID          string                     `json:"authorityId"`
	AuthorityType        string                     `json:"authorityType"`
	OrganizationID       string                     `json:"organizationId"`
	Jurisdiction         *string                    `json:"jurisdiction"`
	Scope                []string                   `json:"scope"`
	VotingPolicy         string                     `json:"votingPolicy"`
	ThresholdClass       string                     `json:"thresholdClass"`
	Threshold            int                        `json:"threshold"`
	EffectiveFromHeight  int64                      `json:"effectiveFromHeight"`
	EffectiveUntilHeight *int64                     `json:"effectiveUntilHeight"`
	Status               string                     `json:"status"`
	Members              []ProtocolGovernanceMember `json:"members"`
}

type ProtocolState struct {
	Profile                string `json:"profile"`
	ChainID                string `json:"chainId"`
	ProtocolVersion        string `json:"protocolVersion"`
	CanonicalSchemaVersion string `json:"canonicalSchemaVersion"`
	ConsensusRulesHash     string `json:"consensusRulesHash"`
	ResourcePolicyHash     string `json:"resourcePolicyHash"`
	ModuleRegistryHash     string `json:"moduleRegistryHash"`
	ActiveFromHeight       int64  `json:"activeFromHeight"`
}

type ProtocolChangeManifest struct {
	ManifestVersion          string `json:"manifestVersion"`
	ChainID                  string `json:"chainId"`
	FromProtocolVersion      string `json:"fromProtocolVersion"`
	ToProtocolVersion        string `json:"toProtocolVersion"`
	RationaleHash            string `json:"rationaleHash"`
	SecurityImpactHash       string `json:"securityImpactHash"`
	CompatibilityImpactHash  string `json:"compatibilityImpactHash"`
	MigrationPlanHash        string `json:"migrationPlanHash"`
	TestEvidenceRoot         string `json:"testEvidenceRoot"`
	MinimumValidatorVersion  string `json:"minimumValidatorVersion"`
	ResourceProfileHash      string `json:"resourceProfileHash"`
}

type ProtocolChangeAction struct {
	Domain                    string `json:"domain"`
	ActionVersion             string `json:"actionVersion"`
	ChainID                   string `json:"chainId"`
	ActionID                  string `json:"actionId"`
	FromProtocolVersion       string `json:"fromProtocolVersion"`
	ToProtocolVersion         string `json:"toProtocolVersion"`
	ApprovedAtHeight          int64  `json:"approvedAtHeight"`
	ActivationHeight          int64  `json:"activationHeight"`
	PreviousProtocolStateHash string `json:"previousProtocolStateHash"`
	NextProtocolStateHash     string `json:"nextProtocolStateHash"`
	GovernancePolicyHash      string `json:"governancePolicyHash"`
	ChangeHash                string `json:"changeHash"`
}

type ProtocolChangeSignature struct {
	AuthorityMemberID string `json:"authorityMemberId"`
	KeyID             string `json:"keyId"`
	Algorithm         string `json:"algorithm"`
	Domain            string `json:"domain"`
	SignatureB64      string `json:"signatureB64"`
}

type ProtocolChangeProof struct {
	Action                ProtocolChangeAction      `json:"action"`
	GovernancePolicy      ProtocolGovernancePolicy  `json:"governancePolicy"`
	ChangeManifest        ProtocolChangeManifest    `json:"changeManifest"`
	PreviousProtocolState ProtocolState             `json:"previousProtocolState"`
	NextProtocolState     ProtocolState             `json:"nextProtocolState"`
	Signatures            []ProtocolChangeSignature `json:"signatures"`
}

type ProtocolChangeVerification struct {
	Valid                     bool     `json:"valid"`
	ActionID                  string   `json:"actionId"`
	ApprovedAtHeight          int64    `json:"approvedAtHeight"`
	ActivationHeight          int64    `json:"activationHeight"`
	FromProtocolVersion       string   `json:"fromProtocolVersion"`
	ToProtocolVersion         string   `json:"toProtocolVersion"`
	GovernancePolicyHash      string   `json:"governancePolicyHash"`
	ChangeHash                string   `json:"changeHash"`
	PreviousProtocolStateHash string   `json:"previousProtocolStateHash"`
	NextProtocolStateHash     string   `json:"nextProtocolStateHash"`
	EligibleAuthorityCount    int      `json:"eligibleAuthorityCount"`
	RequiredAuthority         int      `json:"requiredAuthority"`
	ValidSigners              []string `json:"validSigners"`
}

func protocolGovernanceMemberEligibleAtHeight(member ProtocolGovernanceMember,height int64)bool{return member.State=="ACTIVE"&&member.ValidFromHeight<=height&&(member.ValidUntilHeight==nil||*member.ValidUntilHeight>height)}

func canonicalProtocolGovernancePolicyHash(policy ProtocolGovernancePolicy)(string,error){
	copyPolicy:=policy;copyPolicy.Scope=append([]string(nil),policy.Scope...);sort.Strings(copyPolicy.Scope);copyPolicy.Members=append([]ProtocolGovernanceMember(nil),policy.Members...)
	sort.Slice(copyPolicy.Members,func(i,j int)bool{if copyPolicy.Members[i].AuthorityMemberID==copyPolicy.Members[j].AuthorityMemberID{return copyPolicy.Members[i].KeyID<copyPolicy.Members[j].KeyID};return copyPolicy.Members[i].AuthorityMemberID<copyPolicy.Members[j].AuthorityMemberID})
	return canonicalHashValue(copyPolicy)
}

func requiredProtocolAuthority(policy ProtocolGovernancePolicy,n int)(int,error){
	if n<1{return 0,errors.New("protocol governance requires at least one eligible authority member")}
	floor:=1
	switch policy.ThresholdClass{case "DUAL","MULTI_PARTY_HIGH_ASSURANCE":floor=2;case "MAJORITY":floor=n/2+1;case "SUPERMAJORITY":floor=(2*n)/3+1;case "SIMPLE":default:return 0,fmt.Errorf("unsupported protocol threshold class %s",policy.ThresholdClass)}
	if policy.Threshold>floor{floor=policy.Threshold};return floor,nil
}

func protocolPolicyHasScope(policy ProtocolGovernancePolicy,scope string)bool{for _,candidate:=range policy.Scope{if candidate==scope{return true}};return false}

func verifyProtocolChangeProof(proof ProtocolChangeProof,expectedChainID,expectedGovernancePolicyHash,expectedPreviousProtocolStateHash string)(ProtocolChangeVerification,error){
	action:=proof.Action;policy:=proof.GovernancePolicy;manifest:=proof.ChangeManifest;previous:=proof.PreviousProtocolState;next:=proof.NextProtocolState
	if action.Domain!=protocolChangeDomain||action.ActionVersion!=protocolChangeVersion{return ProtocolChangeVerification{},errors.New("unsupported protocol-change domain/version")}
	if policy.Profile!=protocolGovernanceProfile||policy.AuthorityType!="PROTOCOL"||policy.VotingPolicy!="THRESHOLD"{return ProtocolChangeVerification{},errors.New("unsupported PROTOCOL governance policy profile")}
	if manifest.ManifestVersion!=protocolChangeManifestVersion||previous.Profile!=protocolStateProfile||next.Profile!=protocolStateProfile{return ProtocolChangeVerification{},errors.New("unsupported protocol manifest/state profile")}
	if action.ChainID!=expectedChainID||policy.ChainID!=expectedChainID||manifest.ChainID!=expectedChainID||previous.ChainID!=expectedChainID||next.ChainID!=expectedChainID{return ProtocolChangeVerification{},errors.New("protocol governance chainId mismatch")}
	if action.ActivationHeight<=action.ApprovedAtHeight{return ProtocolChangeVerification{},errors.New("consensus-affecting protocol change must activate after approval height")}
	if policy.Status!="ACTIVE"||policy.EffectiveFromHeight>action.ApprovedAtHeight||(policy.EffectiveUntilHeight!=nil&&*policy.EffectiveUntilHeight<=action.ApprovedAtHeight){return ProtocolChangeVerification{},errors.New("PROTOCOL governance policy was not ACTIVE at approval height")}
	policyHash,err:=canonicalProtocolGovernancePolicyHash(policy);if err!=nil{return ProtocolChangeVerification{},err};if policyHash!=expectedGovernancePolicyHash||action.GovernancePolicyHash!=policyHash{return ProtocolChangeVerification{},errors.New("protocol governance policy hash does not match independently trusted policy")}
	if action.FromProtocolVersion!=previous.ProtocolVersion||action.ToProtocolVersion!=next.ProtocolVersion||manifest.FromProtocolVersion!=previous.ProtocolVersion||manifest.ToProtocolVersion!=next.ProtocolVersion{return ProtocolChangeVerification{},errors.New("protocol version transition is not consistently bound")}
	if previous.ProtocolVersion==next.ProtocolVersion{return ProtocolChangeVerification{},errors.New("protocol activation does not change protocolVersion")}
	if previous.ActiveFromHeight>action.ActivationHeight-1{return ProtocolChangeVerification{},errors.New("previous protocol state was not active before transition height")}
	if next.ActiveFromHeight!=action.ActivationHeight{return ProtocolChangeVerification{},errors.New("next protocol state must activate exactly at governed activationHeight")}
	previousHash,err:=canonicalHashValue(previous);if err!=nil{return ProtocolChangeVerification{},err};nextHash,err:=canonicalHashValue(next);if err!=nil{return ProtocolChangeVerification{},err}
	if previousHash!=expectedPreviousProtocolStateHash||action.PreviousProtocolStateHash!=previousHash{return ProtocolChangeVerification{},errors.New("previous protocol state hash does not match independently trusted state")};if action.NextProtocolStateHash!=nextHash{return ProtocolChangeVerification{},errors.New("next protocol state hash mismatch")}
	changeHash,err:=canonicalHashValue(manifest);if err!=nil{return ProtocolChangeVerification{},err};if action.ChangeHash!=changeHash{return ProtocolChangeVerification{},errors.New("protocol change manifest hash mismatch")}
	if !protocolPolicyHasScope(policy,"PROTOCOL_VERSION"){return ProtocolChangeVerification{},errors.New("protocol governance policy lacks PROTOCOL_VERSION authority")};if previous.ConsensusRulesHash!=next.ConsensusRulesHash&&!protocolPolicyHasScope(policy,"CONSENSUS_RULES"){return ProtocolChangeVerification{},errors.New("protocol governance policy lacks CONSENSUS_RULES authority")};if previous.CanonicalSchemaVersion!=next.CanonicalSchemaVersion&&!protocolPolicyHasScope(policy,"CANONICAL_SCHEMA"){return ProtocolChangeVerification{},errors.New("protocol governance policy lacks CANONICAL_SCHEMA authority")};if previous.ResourcePolicyHash!=next.ResourcePolicyHash&&!protocolPolicyHasScope(policy,"RESOURCE_POLICY"){return ProtocolChangeVerification{},errors.New("protocol governance policy lacks RESOURCE_POLICY authority")}
	for label,value:=range map[string]string{"governancePolicyHash":action.GovernancePolicyHash,"changeHash":action.ChangeHash,"previousProtocolStateHash":action.PreviousProtocolStateHash,"nextProtocolStateHash":action.NextProtocolStateHash,"rationaleHash":manifest.RationaleHash,"securityImpactHash":manifest.SecurityImpactHash,"compatibilityImpactHash":manifest.CompatibilityImpactHash,"migrationPlanHash":manifest.MigrationPlanHash,"testEvidenceRoot":manifest.TestEvidenceRoot,"resourceProfileHash":manifest.ResourceProfileHash,"consensusRulesHash":next.ConsensusRulesHash,"resourcePolicyHash":next.ResourcePolicyHash,"moduleRegistryHash":next.ModuleRegistryHash}{if !isSHA256(value){return ProtocolChangeVerification{},fmt.Errorf("%s must be lowercase SHA-256",label)}}
	eligible:=map[string]ProtocolGovernanceMember{};for _,member:=range policy.Members{if protocolGovernanceMemberEligibleAtHeight(member,action.ApprovedAtHeight){if _,exists:=eligible[member.AuthorityMemberID];exists{return ProtocolChangeVerification{},fmt.Errorf("duplicate protocol authority member %s",member.AuthorityMemberID)};eligible[member.AuthorityMemberID]=member}}
	required,err:=requiredProtocolAuthority(policy,len(eligible));if err!=nil{return ProtocolChangeVerification{},err};if required>len(eligible){return ProtocolChangeVerification{},fmt.Errorf("protocol governance threshold %d exceeds %d eligible authority members",required,len(eligible))}
	normalizedAction,err:=canonicalValue(action);if err!=nil{return ProtocolChangeVerification{},err};canonicalAction,err:=canonicalJSON(normalizedAction);if err!=nil{return ProtocolChangeVerification{},err};payload:=[]byte(canonicalAction)
	valid:=map[string]bool{};for _,sig:=range proof.Signatures{if valid[sig.AuthorityMemberID]{continue};member,ok:=eligible[sig.AuthorityMemberID];if !ok||sig.KeyID!=member.KeyID||sig.Algorithm!="Ed25519"||sig.Domain!=protocolChangeDomain{continue};der,err:=base64.StdEncoding.DecodeString(member.PublicKeyDerB64);if err!=nil{continue};parsed,err:=x509.ParsePKIXPublicKey(der);if err!=nil{continue};pub,ok:=parsed.(ed25519.PublicKey);if !ok{continue};signature,err:=base64.StdEncoding.DecodeString(sig.SignatureB64);if err!=nil{continue};if ed25519.Verify(pub,payload,signature){valid[sig.AuthorityMemberID]=true}}
	validSigners:=make([]string,0,len(valid));for signer:=range valid{validSigners=append(validSigners,signer)};sort.Strings(validSigners);if len(validSigners)<required{return ProtocolChangeVerification{},fmt.Errorf("protocol governance threshold not met: %d/%d; %d required",len(validSigners),len(eligible),required)}
	return ProtocolChangeVerification{true,action.ActionID,action.ApprovedAtHeight,action.ActivationHeight,previous.ProtocolVersion,next.ProtocolVersion,policyHash,changeHash,previousHash,nextHash,len(eligible),required,validSigners},nil
}
